File: solutions/linked-list/python/0141-linked-list-cycle.py
"""
# Difficulty: Medium

# 0141. Linked List Cycle

Given head, the head of a linked list, determine if the linked list has a cycle in it.

There is a cycle in a linked list if there is some node in the list that can be reached again by continuously following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to. Note that pos is not passed as a parameter.

Return true if there is a cycle in the linked list. Otherwise, return false.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[3,2,0,-4], pos = 1 (cycle at node 2)</dd>
<dt>Output:</dt>
<dd>True (has cycle)</dd>
<dt>Explanation:</dt>
<dd>There is a cycle in the linked list where the tail connects back to the second node (index 1)</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>
### METADATA:
**Techniques**: Floyd's Cycle Detection (Tortoise and Hare), Two Pointers
**Data Structures**: Linked List
**Patterns**: Fast and Slow Pointers, Cycle Detection
**Time Complexity**: **O(n)** - Fast pointer traverses at most 2n nodes
**Space Complexity**: **O(1)** - Only two pointer variables used

### INTUITION:
The key insight is that use Floyd's cycle detection with fast and slow pointers. Fast moves 2 steps, slow moves 1 step. If they meet, cycle exists. If fast reaches null, no cycle.

### APPROACH:
1. **Initialize two pointers**: Set slow = fast = head
2. **Move at different speeds**: In loop, slow moves 1 step, fast moves 2 steps
3. **Check for cycle**: If slow == fast, cycle detected, return True
4. **Check for end**: If fast or fast.next is None, no cycle exists
5. **Continue until resolved**: Keep moving pointers
6. **Return False**: If loop exits without meeting, return False

### WHY THIS WORKS:
- This ensures that floyd's tortoise and hare algorithm: if there's a cycle, fast pointer will eventually lap slow pointer
- This ensures that fast moves 2x speed, so it closes gap by 1 node per iteration, guaranteed to meet
- This ensures that if fast reaches null, no cycle exists (linear structure)
- This ensures that o(n) time: worst case fast travels 2n nodes before meeting or reaching null
- This ensures that o(1) space: only two pointers regardless of list size

### EXAMPLE WALKTHROUGH:
Input:
```
head = [3,2,0,-4], pos = 1 (cycle at node 2)
```

Step 1: Fast and slow pointers
slow=3, fast=3
slow=2, fast=0
slow=0, fast=2

Steps:
Step 1: slow=-4, fast=-4 → meet

Output:
```
True (has cycle)
```

### TIME COMPLEXITY:
**O(n)** where n is the number of nodes in the linked list. In the worst case (no cycle), the fast pointer traverses the entire list once, visiting at most n nodes. In the case of a cycle, the fast pointer will meet the slow pointer within the cycle. The meeting happens within n steps because the fast pointer closes the gap by 1 node per iteration. Therefore, maximum operations = 2n (fast pointer moves) = O(n).

### SPACE COMPLEXITY:
**O(1)** - We use only two pointer variables (slow and fast) regardless of the input size. No additional data structures like hash sets or arrays are allocated. The space used doesn't grow with the number of nodes in the list, giving us constant O(1) space complexity. This is more space-efficient than the hash set approach which would require O(n) space.

### EDGE CASES:
- **Empty list**: head = None returns False (no cycle possible)
- **Single node no cycle**: head = [1], pos = -1 returns False
- **Single node with cycle**: head = [1], pos = 0 returns True (points to itself)
- **Two nodes with cycle**: head = [1,2], pos = 0 returns True (forms cycle)
- **Large list no cycle**: Returns False after traversing entire list
- **Cycle at end**: Tail connects back to any previous node, detected correctly
- **Very long cycle**: Fast pointer eventually catches slow pointer within cycle

</details>
"""

from typing import Any


class ListNode:
    def __init__(self, val: Any = 0, next: Any = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        """
        Detect if linked list has a cycle using Floyd's cycle detection algorithm.

        Args:
            head: Head of the linked list

        Returns:
            True if cycle exists, False otherwise

        Time Complexity: O(n) - fast pointer traverses at most 2n nodes
        Space Complexity: O(1) - only using two pointers
        """
        if not head or not head.next:
            return False

        # Initialize slow and fast pointers
        slow = head
        fast = head.next

        # Floyd's cycle detection (tortoise and hare)
        while slow != fast:
            if not fast or not fast.next:
                return False
            slow = slow.next
            fast = fast.next.next

        return True

    def solve(self, head: ListNode) -> bool:
        """Wrapper method for consistency with template."""
        return self.hasCycle(head)


def create_cycle_list(values: list[Any], pos: int) -> ListNode:
    """
    Create a linked list with a cycle.

    Args:
        values: List of node values
        pos: Position where tail connects (-1 for no cycle)

    Returns:
        Head of the linked list
    """
    if not values:
        return None  # type: ignore

    head = ListNode(values[0])
    current = head
    cycle_node = head if pos == 0 else None

    for i, val in enumerate(values[1:], 1):
        current.next = ListNode(val)
        current = current.next
        if i == pos:
            cycle_node = current

    # Create cycle if pos >= 0
    if pos >= 0 and cycle_node:
        current.next = cycle_node

    return head


def test_solution() -> None:
    """
    Test cases for 141. Linked List Cycle.
    """
    solution = Solution()

    # Test case 1: List with cycle
    head = create_cycle_list([3, 2, 0, -4], 1)
    solution.solve(head)
    # # # assert result == expected, f"Expected expected, got result"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined

    # Test case 2: List with cycle at head
    head = create_cycle_list([1, 2], 0)
    solution.solve(head)
    # # # assert result == expected, f"Expected expected, got result"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined

    # Test case 3: No cycle
    head = create_cycle_list([1, 2, 3, 4], -1)
    solution.solve(head)
    # # # assert result == expected, f"Expected expected, got result"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined

    # Test case 4: Single node, no cycle
    head = create_cycle_list([1], -1)
    solution.solve(head)
    # # # assert result == expected, f"Expected expected, got result"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined

    # Test case 5: Empty list
    head = None  # type: ignore
    solution.solve(head)
    # # # assert result == expected, f"Expected expected, got result"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    head = create_cycle_list([3, 2, 0, -4], 1)
    solution.solve(head)

File: solutions/linked-list/typescript/0141-linked-list-cycle.ts
/**
 * ### METADATA:
 *
 *
 * ### INTUITION:
 * The key insight is that use Floyd's cycle detection with fast and slow pointers. Fast moves 2 steps, slow moves 1 step. If they meet, cycle exists. If fast reaches null, no cycle.
 *
 * ### APPROACH:
 * 1. **Initialize two pointers**: Set slow = fast = head
 * 2. **Move at different speeds**: In loop, slow moves 1 step, fast moves 2 steps
 * 3. **Check for cycle**: If slow == fast, cycle detected, return True
 * 4. **Check for end**: If fast or fast.next is None, no cycle exists
 * 5. **Continue until resolved**: Keep moving pointers
 * 6. **Return False**: If loop exits without meeting, return False
 *
 * ### WHY THIS WORKS:
 * - This ensures that floyd's tortoise and hare algorithm: if there's a cycle, fast pointer will eventually lap slow pointer
 * - This ensures that fast moves 2x speed, so it closes gap by 1 node per iteration, guaranteed to meet
 * - This ensures that if fast reaches null, no cycle exists (linear structure)
 * - This ensures that o(n) time: worst case fast travels 2n nodes before meeting or reaching null
 * - This ensures that o(1) space: only two pointers regardless of list size
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * head = [3,2,0,-4], pos = 1 (cycle at node 2)
 * ```
 *
 * Step 1: Fast and slow pointers
 * slow=3, fast=3
 * slow=2, fast=0
 * slow=0, fast=2
 *
 * Steps:
 * Step 1: slow=-4, fast=-4 → meet
 *
 * Output:
 * ```
 * True (has cycle)
 * ```
 *
 * ### TIME COMPLEXITY:
 * O(n)**
 * - Single pass through input
 *
 * ### SPACE COMPLEXITY:
 * **O(n)** - [Explanation of why this complexity]. The algorithm [describe the operation] which takes **O(n)** space.
 *
 * ### EDGE CASES:
 * - **Empty input**: Handle when input is empty
 * - **Single element**: Handle single-element inputs
 * - **Boundary values**: Handle minimum/maximum valid values
 *
 * *
 */

class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

class Solution {
  /**
   * Detect if linked list has a cycle using Floyd's cycle detection algorithm.
   *
   * Time Complexity: O(n)
   * Space Complexity: O(1)
   */
  hasCycle(head: ListNode | null): boolean {
    if (!head || !head.next) {
      return false;
    }

    // Initialize slow and fast pointers
    let slow: ListNode | null = head;
    let fast: ListNode | null = head.next;

    // Floyd's cycle detection (tortoise and hare)
    while (slow !== fast) {
      if (!fast || !fast.next) {
        return false;
      }
      slow = slow!.next;
      fast = fast.next.next;
    }

    return true;
  }

  /**
   * Alternative: using hash set to track visited nodes.
   *
   * Time Complexity: O(n)
   * Space Complexity: O(n)
   */
  hasCycleHashSet(head: ListNode | null): boolean {
    const visited = new Set<ListNode>();
    let current = head;

    while (current) {
      if (visited.has(current)) {
        return true;
      }
      visited.add(current);
      current = current.next;
    }

    return false;
  }
}

/**
 * Helper function to create a linked list with optional cycle.
 */
function createCycleList(values: number[], pos: number): ListNode | null {
  if (values.length === 0) {
    return null;
  }

  const head = new ListNode(values[0]);
  let current = head;
  let cycleNode: ListNode | null = pos === 0 ? head : null;

  for (let i = 1; i < values.length; i++) {
    current.next = new ListNode(values[i]);
    current = current.next;
    if (i === pos) {
      cycleNode = current;
    }
  }

  // Create cycle if pos >= 0
  if (pos >= 0 && cycleNode) {
    current.next = cycleNode;
  }

  return head;
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = { Solution, ListNode };
}

function runTests(): void {
  const solution = new Solution();

  // Test case 1: List with cycle at position 1
  const head1 = createCycleList([3, 2, 0, -4], 1);
  console.log(`Test 1: ${solution.hasCycle(head1) === true ? "PASS" : "FAIL"}`);

  // Test case 2: List with cycle at head
  const head2 = createCycleList([1, 2], 0);
  console.log(`Test 2: ${solution.hasCycle(head2) === true ? "PASS" : "FAIL"}`);

  // Test case 3: No cycle
  const head3 = createCycleList([1, 2, 3, 4], -1);
  console.log(`Test 3: ${solution.hasCycle(head3) === false ? "PASS" : "FAIL"}`);

  // Test case 4: Single node, no cycle
  const head4 = createCycleList([1], -1);
  console.log(`Test 4: ${solution.hasCycle(head4) === false ? "PASS" : "FAIL"}`);

  // Test case 5: Empty list
  const head5 = null;
  console.log(`Test 5: ${solution.hasCycle(head5) === false ? "PASS" : "FAIL"}`);

  // Test case 6: Hash set approach
  const head6 = createCycleList([1, 2, 3], 1);
  console.log(`Test 6: ${solution.hasCycleHashSet(head6) === true ? "PASS" : "FAIL"}`);

  console.log("\nAll test cases completed!");
}

if (typeof require !== "undefined" && require.main === module) {
  runTests();
}

export default Solution;
